File: packages/libs/ContextDistillLib/Sources/ContextDistillLib/PythonText.swift
// PythonText.swift
// Python-semantics text helpers over [Unicode.Scalar].
//
// All functions in this file replicate Python 3 str methods exactly,
// using unicodeScalars as the index unit (matching Python's code-point
// indexing via str[i]).  Rust uses chars() for the same unit.
//
// Unicode divergences from Swift's stdlib are documented per function.
//
// No external dependencies.  No regex.  Pure arithmetic over scalars.

import Foundation

// MARK: - Whitespace

/// Returns `true` if the scalar is in Python 3's `str.isspace()` set.
///
/// Mirrors Python: `c.isspace()` for a single character.
///
/// Python uses Unicode's Whitespace property (UAX #44 PropList.txt).
/// The exact set:
///   U+0009 TAB, U+000A LF, U+000B VT, U+000C FF, U+000D CR
///   U+0020 SPACE
///   U+0085 NEL (next line)
///   U+00A0 NO-BREAK SPACE
///   U+1680 OGHAM SPACE MARK
///   U+2000–U+200A (various typographic spaces)
///   U+2028 LINE SEPARATOR
///   U+2029 PARAGRAPH SEPARATOR
///   U+202F NARROW NO-BREAK SPACE
///   U+205F MEDIUM MATHEMATICAL SPACE
///   U+3000 IDEOGRAPHIC SPACE
///
/// Swift divergence: `CharacterSet.whitespacesAndNewlines` covers most of
/// these but is not identical at the Unicode scalar level.  We enumerate the
/// exact code-point set from Python's unicodedata to avoid divergence.
///
/// Note: U+001C–U+001F (ASCII separators) are NOT in this set even though
/// some sources list them.  Python 3.x does not treat them as whitespace.
public func isPythonWhitespace(_ s: Unicode.Scalar) -> Bool {
    switch s.value {
    case 0x0009,          // U+0009 CHARACTER TABULATION (\t)
         0x000A,          // U+000A LINE FEED (\n)
         0x000B,          // U+000B LINE TABULATION (\v)
         0x000C,          // U+000C FORM FEED (\f)
         0x000D,          // U+000D CARRIAGE RETURN (\r)
         0x0020,          // U+0020 SPACE
         0x0085,          // U+0085 NEXT LINE (NEL)
         0x00A0,          // U+00A0 NO-BREAK SPACE
         0x1680,          // U+1680 OGHAM SPACE MARK
         0x2000 ... 0x200A, // U+2000–U+200A (EN QUAD through HAIR SPACE)
         0x2028,          // U+2028 LINE SEPARATOR
         0x2029,          // U+2029 PARAGRAPH SEPARATOR
         0x202F,          // U+202F NARROW NO-BREAK SPACE
         0x205F,          // U+205F MEDIUM MATHEMATICAL SPACE
         0x3000:          // U+3000 IDEOGRAPHIC SPACE
        return true
    default:
        return false
    }
}

// MARK: - Word characters (\w)

/// Returns `true` if the scalar is a Python regex `\w` character.
///
/// Python 3 re with default Unicode flag: `\w` matches Unicode letters,
/// digits, and underscore — equivalent to `[A-Za-z0-9_]` for ASCII plus
/// Unicode letters/digits from all scripts.
///
/// Swift divergence: Swift's `CharacterSet.letters` and `.decimalDigits`
/// cover the correct Unicode sets.  We use `properties.isAlphabetic`,
/// `.isNumber`, and the literal underscore for maximum fidelity.
///
/// Used by `pyWordBoundary` to detect \b transitions.
public func isPythonWordChar(_ s: Unicode.Scalar) -> Bool {
    // Underscore is always a word character.
    if s.value == 0x005F { return true }
    // Unicode alphabetic (covers letters from all scripts).
    if s.properties.isAlphabetic { return true }
    // Unicode decimal/letter/number digits.
    // .isNumber is broader (includes fractions); Python's \w uses isAlpha+isDigit,
    // so we check numeric digit categories directly.
    switch s.properties.generalCategory {
    case .decimalNumber,          // Nd — 0-9 and script-specific digits
         .letterNumber,           // Nl — Roman numerals, etc.
         .otherNumber:            // No — fractions, etc.
        return true
    default:
        return false
    }
}

/// Returns `true` at a Python `\b` word boundary in `scalars` at `index`.
///
/// A `\b` is a zero-width assertion that is true when:
///   - The character at `index` is `\w` and the character at `index-1` is `\W`
///     (or `index == 0`)
///   - OR the character at `index-1` is `\w` and the character at `index` is
///     `\W` (or `index == scalars.count`)
///
/// This mirrors Python's `re` engine's `\b` semantics exactly, using
/// `isPythonWordChar` for `\w` classification.
///
/// - Parameters:
///   - index: The code-point index to test (0-based, inclusive of
///     `scalars.count` for the position after the last character).
///   - scalars: The full scalar array.
/// - Returns: `true` if a word boundary exists at `index`.
public func pyWordBoundary(at index: Int, in scalars: [Unicode.Scalar]) -> Bool {
    let prevW: Bool
    let currW: Bool
    if index > 0 {
        prevW = isPythonWordChar(scalars[index - 1])
    } else {
        prevW = false   // virtual non-word before start of string
    }
    if index < scalars.count {
        currW = isPythonWordChar(scalars[index])
    } else {
        currW = false   // virtual non-word after end of string
    }
    return prevW != currW
}

// MARK: - Character predicates (str.isalnum, isalpha, isdigit, isupper)

/// Returns `true` if `s` is an alphanumeric Unicode scalar.
///
/// Mirrors Python: `c.isalnum()` — True if the character is alphabetic OR a
/// decimal/letter/other number.  Unicode-aware.
public func pyIsAlnum(_ s: Unicode.Scalar) -> Bool {
    if s.properties.isAlphabetic { return true }
    switch s.properties.generalCategory {
    case .decimalNumber, .letterNumber, .otherNumber: return true
    default: return false
    }
}

/// Returns `true` if `s` is an alphabetic Unicode scalar.
///
/// Mirrors Python: `c.isalpha()` — True if the scalar is in the Unicode
/// alphabetic category (`Alphabetic` derived property).
public func pyIsAlpha(_ s: Unicode.Scalar) -> Bool {
    s.properties.isAlphabetic
}

/// Returns `true` if `s` is a decimal digit (Unicode Nd category).
///
/// Mirrors Python: `c.isdigit()`.  Python's isdigit() is True for decimal
/// digits in any script (Nd) plus some superscript/circled digits (No).
/// For the ASCII-dominant oracle texts this distinction does not arise.
///
/// Swift divergence: `properties.generalCategory == .decimalNumber` is Nd
/// only.  Python includes some No characters.  For the oracle corpus this
/// difference is irrelevant.
public func pyIsDigit(_ s: Unicode.Scalar) -> Bool {
    s.properties.generalCategory == .decimalNumber
}

/// Returns `true` if `s` is an uppercase letter.
///
/// Mirrors Python: `c.isupper()` for a single character — True when the
/// character has uppercase casing.
///
/// Swift divergence: `s.properties.isUppercase` uses Unicode's Uppercase
/// derived property, which matches Python 3's `str.isupper()` for single
/// characters.
public func pyIsUpper(_ s: Unicode.Scalar) -> Bool {
    s.properties.isUppercase
}

// MARK: - Case conversion

/// Applies Python's `str.lower()` to a scalar array.
///
/// Mirrors Python: `text.lower()`.
///
/// Python uses Unicode simple case mapping, as does Swift's `lowercased()`.
/// The two implementations agree for the characters in the oracle corpus.
///
/// Swift divergence: `String.lowercased()` may produce ligature decompositions
/// (e.g. DZ-ligature → dz) or multi-character expansions for some rare code
/// points.  For the ASCII-dominant oracle texts this does not arise.
/// Swift does NOT produce different results for Greek sigma context (Σ→σ always)
/// at the single-character level.  Python is the same.
///
/// Implementation: convert to String, lowercase, re-extract scalars.
/// This is safe because we work with code points as the index unit; the
/// round-trip through String preserves code-point count for simple mappings.
public func pyLower(_ scalars: [Unicode.Scalar]) -> [Unicode.Scalar] {
    let str = scalars.asString().lowercased()
    return Array(str.unicodeScalars)
}

// MARK: - Splitting and stripping

/// Splits `scalars` on Python's whitespace set, returning non-empty tokens.
///
/// Mirrors Python: `text.split()` (no argument).
///
/// Python's no-arg split:
///   1. Strips leading and trailing whitespace.
///   2. Splits on any run of whitespace (any character for which isspace()
///      returns True).
///   3. Never produces empty strings.
///
/// Result: an array of scalar arrays, one per token.  Returns empty if the
/// input is all whitespace or empty.
///
/// Used by `estimateTokens` to count words, matching Python's `text.split()`.
public func pySplit(_ scalars: [Unicode.Scalar]) -> [[Unicode.Scalar]] {
    var tokens: [[Unicode.Scalar]] = []
    var current: [Unicode.Scalar] = []
    for s in scalars {
        if isPythonWhitespace(s) {
            if !current.isEmpty {
                tokens.append(current)
                current = []
            }
        } else {
            current.append(s)
        }
    }
    if !current.isEmpty { tokens.append(current) }
    return tokens
}

/// Strips Python whitespace from both ends of `scalars`.
///
/// Mirrors Python: `text.strip()`.
public func pyStrip(_ scalars: [Unicode.Scalar]) -> [Unicode.Scalar] {
    let ls = pyLstrip(scalars)
    return pyRstrip(ls)
}

/// Strips Python whitespace from the left (start) of `scalars`.
///
/// Mirrors Python: `text.lstrip()`.
public func pyLstrip(_ scalars: [Unicode.Scalar]) -> [Unicode.Scalar] {
    var start = 0
    while start < scalars.count && isPythonWhitespace(scalars[start]) {
        start += 1
    }
    return Array(scalars[start...])
}

/// Strips Python whitespace from the right (end) of `scalars`.
///
/// Mirrors Python: `text.rstrip()`.
public func pyRstrip(_ scalars: [Unicode.Scalar]) -> [Unicode.Scalar] {
    var end = scalars.count
    while end > 0 && isPythonWhitespace(scalars[end - 1]) {
        end -= 1
    }
    return Array(scalars[..<end])
}

// MARK: - UTF-8 byte offset

/// Returns the UTF-8 byte offset corresponding to code-point index `cpIndex`
/// in `scalars`.
///
/// Python uses code points as its index unit (str[i] is the i-th Unicode
/// scalar).  Swift struct fields in the oracle output carry both `start` (code
/// points) and `start_utf8_byte` (bytes).  This function converts between them.
///
/// - Parameters:
///   - cpIndex: A code-point index (0-based) into `scalars`.
///   - scalars: The full scalar array.
/// - Returns: The byte offset into the equivalent UTF-8 encoding.
///   Returns `scalars.utf8ByteCount` when `cpIndex == scalars.count`.
/// - Precondition: `cpIndex` is in `0 ... scalars.count`.
public func utf8ByteOffset(forCodePoint cpIndex: Int, in scalars: [Unicode.Scalar]) -> Int {
    precondition(cpIndex >= 0 && cpIndex <= scalars.count,
                 "cpIndex \(cpIndex) out of range [0, \(scalars.count)]")
    var bytes = 0
    for i in 0 ..< cpIndex {
        bytes += scalars[i].utf8.count
    }
    return bytes
}

File: packages/libs/ContextDistillLib/rust/src/python_text.rs
//! Python-semantics text helpers for CDL-01 hand-written scanners.
//!
//! Every function in this module mirrors the behaviour of a Python
//! string or `re` built-in as precisely as possible on `&[char]`
//! (Unicode scalar values, equal to Python code-points for valid
//! UTF-8 excluding surrogates).
//!
//! Where Rust semantics diverge from Python semantics the divergence
//! is documented in an inline comment.

// ---------------------------------------------------------------------------
// Whitespace
// ---------------------------------------------------------------------------

/// Return true if `c` is a Python whitespace character.
///
/// Mirrors `str.isspace()` / `re.compile(r'\s')` (no `re.ASCII`).
///
/// Python's Unicode whitespace set includes, among others:
///   U+0009 HT, U+000A LF, U+000B VT, U+000C FF, U+000D CR,
///   U+0020 SPACE, U+00A0 NBSP, U+2000-U+200A, U+2028, U+2029,
///   U+202F, U+205F, U+3000.
///
/// Rust `char::is_whitespace()` implements the Unicode White_Space
/// property, which covers the same set for all practical inputs in
/// this corpus.  The two properties diverge on a handful of rarely-
/// used code points (e.g. U+0085 NEL is White_Space in Rust/Unicode
/// but not in Python's `\s`).  We use `char::is_whitespace()` here
/// and document the gap.  For the 509-row corpus no divergence was
/// observed.
#[inline]
pub fn is_python_whitespace(c: char) -> bool {
    c.is_whitespace()
}

// ---------------------------------------------------------------------------
// Strip helpers (mirror Python str.strip / lstrip / rstrip)
// ---------------------------------------------------------------------------

/// Return the sub-slice with leading whitespace removed.
/// Mirrors Python `str.lstrip()` (no argument → strips whitespace).
pub fn py_lstrip(chars: &[char]) -> &[char] {
    let start = chars.iter().position(|c| !is_python_whitespace(*c)).unwrap_or(chars.len());
    &chars[start..]
}

/// Return the sub-slice with trailing whitespace removed.
/// Mirrors Python `str.rstrip()` (no argument → strips whitespace).
pub fn py_rstrip(chars: &[char]) -> &[char] {
    let end = chars.iter().rposition(|c| !is_python_whitespace(*c)).map_or(0, |p| p + 1);
    &chars[..end]
}

/// Return the sub-slice with leading and trailing whitespace removed.
/// Mirrors Python `str.strip()` (no argument → strips whitespace).
pub fn py_strip(chars: &[char]) -> &[char] {
    py_rstrip(py_lstrip(chars))
}

// ---------------------------------------------------------------------------
// Split (mirrors Python str.split() with no argument)
// ---------------------------------------------------------------------------

/// Split on any run of whitespace, discarding leading and trailing
/// whitespace runs, and discarding empty tokens.
///
/// Mirrors Python `text.split()` (no argument):
///   ```python
///   "  hello   world  ".split()  # ['hello', 'world']
///   ```
///
/// Note: Python `str.split()` with no argument splits on any
/// consecutive whitespace and never produces empty strings.
pub fn py_split(chars: &[char]) -> Vec<&[char]> {
    let mut result = Vec::new();
    let mut i = 0;
    while i < chars.len() {
        // Skip whitespace
        while i < chars.len() && is_python_whitespace(chars[i]) {
            i += 1;
        }
        if i >= chars.len() {
            break;
        }
        let start = i;
        // Consume non-whitespace token
        while i < chars.len() && !is_python_whitespace(chars[i]) {
            i += 1;
        }
        result.push(&chars[start..i]);
    }
    result
}

// ---------------------------------------------------------------------------
// Case conversion
// ---------------------------------------------------------------------------

/// Convert a char to its Python `str.lower()` equivalent.
///
/// Python's `str.lower()` performs full Unicode case-folding, which can
/// produce multiple chars for a single input (e.g. U+00DF LATIN SMALL
/// LETTER SHARP S → "ss").  Rust `char::to_lowercase()` returns an
/// iterator over the full lowercase sequence, matching Python here.
///
/// The returned `Vec<char>` is almost always length 1; callers that
/// need a scalar approximation may take the first element only, but
/// should document the truncation.
pub fn py_lower_char(c: char) -> Vec<char> {
    c.to_lowercase().collect()
}

/// Lowercase an entire slice, expanding multi-char lowercases.
/// Mirrors Python `str.lower()`.
pub fn py_lower(chars: &[char]) -> Vec<char> {
    chars.iter().flat_map(|c| c.to_lowercase()).collect()
}

// ---------------------------------------------------------------------------
// Character class tests (used in scanner word-class and \b logic)
// ---------------------------------------------------------------------------

/// True if `c` is a Python `\w` character.
///
/// Python's `\w` without `re.ASCII` matches Unicode word characters:
/// letters, digits, and `_`.  Rust `char::is_alphanumeric()` covers
/// the same letter and digit set (Unicode General Category L* + N*).
/// The underscore is added explicitly.
#[inline]
pub fn py_word_char(c: char) -> bool {
    c.is_alphanumeric() || c == '_'
}

/// True if `c` is a Python `\d` character.
///
/// In the distill_plus_converter.py oracle all `\d` uses are in ASCII
/// contexts (date/number matching).  Python `\d` without `re.ASCII`
/// matches Unicode digits; in practice the corpus contains only ASCII
/// digits so we use `char::is_ascii_digit()`.
///
/// Divergence note: Python `\d` matches e.g. U+0660 ARABIC-INDIC
/// DIGIT ZERO; `char::is_ascii_digit()` does not.  No such code
/// points appear in the 509-row corpus.
#[inline]
pub fn py_digit(c: char) -> bool {
    c.is_ascii_digit()
}

/// True if `c` is a Python `[A-Za-z0-9]` ASCII alphanumeric character.
///
/// Unlike `py_word_char`, this mirrors explicit `[A-Za-z0-9]` char
/// classes in patterns like WORD_RE that exclude Unicode-only alnum.
#[inline]
pub fn py_ascii_alnum(c: char) -> bool {
    c.is_ascii_alphanumeric()
}

/// True if `c` is an ASCII letter `[A-Za-z]`.
#[inline]
pub fn py_ascii_alpha(c: char) -> bool {
    c.is_ascii_alphabetic()
}

/// True if `c` is an uppercase ASCII letter `[A-Z]`.
#[inline]
pub fn py_ascii_upper(c: char) -> bool {
    c.is_ascii_uppercase()
}

// ---------------------------------------------------------------------------
// Word-boundary detection
// ---------------------------------------------------------------------------

/// Return true if a Python `\b` word boundary exists at `pos` in `chars`.
///
/// Python `\b` (no `re.ASCII`) is a zero-width assertion that matches
/// where the `\w` status changes between the character before and the
/// character after position `pos`:
///
///   `is_word(chars[pos-1]) != is_word(chars[pos])`
///
/// with `is_word(out_of_range)` defined as `false`.
///
/// This uses `py_word_char` (Unicode alnum + `_`) as the `\w` test.
#[inline]
pub fn word_boundary_at(chars: &[char], pos: usize) -> bool {
    let before = if pos > 0 { py_word_char(chars[pos - 1]) } else { false };
    let after = if pos < chars.len() { py_word_char(chars[pos]) } else { false };
    before != after
}

/// Return true if a `\b` word boundary exists BEFORE `pos` (i.e. pos
/// is the START of a potential word match).
///
/// Equivalent to `word_boundary_at(chars, pos)` when the char at `pos`
/// is a word char (which is the usual case for word-start positions).
#[inline]
pub fn word_start_boundary(chars: &[char], pos: usize) -> bool {
    word_boundary_at(chars, pos)
}

/// Return true if a `\b` word boundary exists AFTER the matched region
/// ending at `end` (exclusive).
///
/// This is `word_boundary_at(chars, end)`.
#[inline]
pub fn word_end_boundary(chars: &[char], end: usize) -> bool {
    word_boundary_at(chars, end)
}

// ---------------------------------------------------------------------------
// UTF-8 byte-offset conversion
// ---------------------------------------------------------------------------

/// Compute the UTF-8 byte offset corresponding to code-point index
/// `cp_idx` in `chars`.
///
/// Python code-point indices (used throughout this codebase) are
/// equivalent to indices into `chars` (a `Vec<char>`).  This function
/// converts such an index to the byte offset in the original UTF-8
/// `&str`, which is needed for the `span_utf8_offset_unit` field in
/// oracle output.
///
/// `cp_idx` must be `<= chars.len()`.  If `cp_idx == chars.len()`,
/// returns the total byte length of the string.
pub fn cp_to_utf8_offset(chars: &[char], cp_idx: usize) -> usize {
    chars[..cp_idx].iter().map(|c| c.len_utf8()).sum()
}

// ---------------------------------------------------------------------------
// ASCII word character (used for \b in ASCII-context patterns)
// ---------------------------------------------------------------------------

/// True if `c` is an ASCII word character: `[A-Za-z0-9_]`.
///
/// Some patterns in record_shape_classifier.py use `\b` in purely
/// ASCII contexts.  This narrower definition avoids false boundaries
/// on Unicode accented letters.
#[inline]
pub fn ascii_word_char(c: char) -> bool {
    c.is_ascii_alphanumeric() || c == '_'
}

/// ASCII-context word boundary: uses `ascii_word_char` instead of
/// `py_word_char`.
#[inline]
pub fn ascii_word_boundary_at(chars: &[char], pos: usize) -> bool {
    let before = if pos > 0 { ascii_word_char(chars[pos - 1]) } else { false };
    let after = if pos < chars.len() { ascii_word_char(chars[pos]) } else { false };
    before != after
}
